util/src/main/java/io/kubernetes/client/Exec.java

```diff
@@ -371,6 +371,8 @@ protected void handleMessage(int stream, InputStream inStream) throws IOExceptio
                   }
                 }
                 inStream.close();
+                // The websocket needs to be closed.
+                this.close();
```

Comment:
and we should call `ExecProcess.this.notifyAll();` in the end, are we?

Reply:
yes, we are but that will just release the locks and resume the blocked threads (if any) but it won't close the WebSocket.
I believe the WebSocket connection is not getting close by itself nor the input stream.
And eventually, the code gets hanged at `ByteStreams.copy(dataStream, System.out);` in `CopyExample`.

Reply:
> yes, we are but that will just release the locks and resume the blocked threads (if any)

and it will improve the robustness. how about releasing a CountDownLatch here?

> I believe the WebSocket connection is not getting close by itself nor the input stream.
And eventually, the code gets hanged at ByteStreams.copy(dataStream, System.out); in CopyExample.

if i'm understanding correctly, as the websocket is a full-duplex communication, there's supposed to be two "close" action for completely shutting down the connection:
- one is at the server-side, triggerred by the k8s apiserver.
- the other is at the client-side, by invoking `socket.close()` somewhere else

so the websocket isn't "closed" as long as either side isn't yet closed. after the k8s apiserver signals an exit code by the pipe id-3 (reaching this line), it should do the server-side close, which on the other hand triggers the client-side close. so theoretically we don't need to close it again upon exit codes. but as the close call looks idempotent, closing it twice wouldn't be risky..

Reply:
> after the k8s apiserver signals an exit code by the pipe id-3 (reaching this line), it should do the server-side close,

Unfortunately, at this point, k8s apiserver is NOT closing the connection from there end but on the other side, if the client itself initiates the WebSocket close, post that k8s apiserver also invokes `close()` method of `WebSocketStreamHandler`.